### DISCUSSION TITLE
Need max redelivery count at message level.

### DISCUSSION THREAD

**Discussion by Raunak-Agrawal** [apache/pulsar#17626] [Q&A]
I am trying to implement retries at the message level. Depending on the type of message ( which is failed due to some exceptions) , I want to have different retry policy. Like some messages will have exponential retry with max retries count of 10. While some messages will have constant retry with retries happening till 12pm midnight. It could also be for eg. Depending on type of exception, I want to retry it differently.

I am trying to use reconsumeLater() to do this. But I am not sure how to set maxRedelivery at message level. Because from the docs, I can see, it needs to be set at consumer level. If we set at consumer level, all messages will inherit the same configuration.

I am also expecting after the retries are exhausted, messages move to the DLQ. 

Can anyone help me on how can I build such retry policy and attach to message.

**Comment by Raunak-Agrawal** [comment]
Can someone please help me with this? @michaeljmarshall

**Comment by Jason918** [comment]
We don't have `maxRedelivery` at message level. Currently, you can get reconsumed times of a message from its properties with key `RetryMessageUtil.SYSTEM_PROPERTY_RECONSUMETIMES` ("RECONSUMETIMES"), and then decide whether call reconsumeLater().

> I am also expecting after the retries are exhausted, messages move to the DLQ.

As for DLQ, it's basically a normal topic. You can create your own, and send messages using your own producer.

**Reply by michaeljmarshall** [reply]
Great points @Jason918. Mixing retry patterns is not currently supported, however you can definitely add this logic to an application using the Pulsar client. Delayed message delivery and a regex topic consumer might help achieve that logic, too.

**Reply by raunakagrawal47** [reply]
One question I have is, If I manually move the message to DLQ after my custom implementation for the number of retries, does that message gets acknowledged/removed from the main queue? or how does it work? How does message gets removed from the main queue? @Jason918 @michaeljmarshall

**Reply by Jason918** [reply]
> How does message gets removed from the main queue?

You can manually ack this message from main queue after you send the message to DLQ successfully.

**Reply by raunakagrawal47** [reply]
Thanks for your reply @Jason918 . One last question, do we need a retry topic to implement reconsumeLater()? Or can it work without that?
So, basically, I have a main topic which is subscribed to main events and then one DLQ topic to send errored messages to. Now, do I also need a retry topic if I am using reconsumeLater() to redeliver messages?

I tried using reconsumeLater() without retry topic and still i was able to get messages redelivered. 
Also, one more question is, if a consumer goes down due to some issues, does the messages get replayed back from the main queue or the retry topic queue?

**Reply by Jason918** [reply]
Have you set `deadLetterPolicy` in `ConsumerBuilder` ?
It must be set before you can call `reconsumeLater()`. And once you do, it will create its own retry topic and DLQ with default settings. You can also define your own DLQ in the deadLetterPolicy.

Basically, you can set `DeadLetterPolicy#maxRedeliverCount` to a very large number, and `DeadLetterPolicy#deadLetterTopic` to your own dead letter topic. Once you decide one message should goes to DLQ, you send it manually with your own DLQ producer.

> Also, one more question is, if a consumer goes down due to some issues, does the messages get replayed back from the main queue or the retry topic queue?

Yes, as long as the message is not acked and not exceeded backlog quota or retention policy. It will get replayed.

**Reply by raunakagrawal47** [reply]
My consumer looks like this right now:

> pulsarClient.newConsumer()
                    .topic("my-topic")
                    .subscriptionType(SubscriptionType.Shared)
                    .subscriptionName("my-subscription")
                    .receiverQueueSize(10)
                    .enableRetry(true)
                    .subscribe()

                    
 I have not set any `deadLetterPolicy ` and `ackTimeout` since I wanted to send messages to DLQ using my own producer after max retries are exhausted. Max retries are also defined at the message level.
 
 Also, I have some logic written for sending the message to DLQ. Since I am sending to DLQ based on expire duration set for each message, I have not put `ackTimeout` at the consumer level.
 
 

>  if (hours > retryExpiryAfterDuration) {   // Don't retry after expiry duration set. Hour is calculated from the time the message first arrived difference from the current time.
                        // send to DLQ .... retries are exhausted.
                        System.out.println("Sending to DLQ after max number of retries");
                        MessageId messageId = pulsarProducerDLQ.send(String.valueOf(msg));
                        if (messageId != null) {
                            consumer.acknowledge(msg); // acknowledge here to remove from the main queue.
                        }
                        return;
                    }    

   
   
   Since DLQ is just another topic, I was able to send a message to the DLQ topic using `pulsarProducerDLQ`. 
   

> this.pulsarClient.newProducer().topic('my-dlq-topic').messageRoutingMode(RoundRobinPartition).
                    compressionType(CompressionType.LZ4).create().newMessage()
                    .key(message)
                    .value(message.getBytes())
                    .send();

   
   Could you please let me know if this is the right way to go about it ?

**Reply by Jason918** [reply]
> .enableRetry(true)

Please notice: This will set the dead letter policy implicitly with `maxRedeliverCount=16`, so if you calls `reconsumeLater` on the same message over 16 times, it will goes to its own default DLQ.

**Reply by raunakagrawal47** [reply]
@Jason918 @michaeljmarshall 
1. If I just set` enableRetry(true)`, how much time does it wait to confirm whether an acknowledgement is done or not and replay the message? Because we are not using any timeout here. 
2. Even If I set `maxRedeliverCount(1000)` since I am manually sending the message to DLQ using my own producer based on max retries using my policy for each message, how will the message get ack'ed/removed from the retry-queue? If i do `consumer.acknowledge(msg);` after sending to DLQ, does it remove from main queue or retry-queue?
3. We are processing millions of requests in a day, does this scale well?

**Reply by Jason918** [reply]
> 1. If I just set` enableRetry(true)`, how much time does it wait to confirm whether an acknowledgement is done or not and replay the message? Because we are not using any timeout here.

IIUC, it won't replay unless the consumer reconnects, e.g. consumer restart or broker restarts.



> * Even If I set `maxRedeliverCount(1000)` since I am manually sending the message to DLQ using my own producer based on max retries using my policy for each message, how will the message get ack'ed/removed from the retry-queue? 

You acked it just like it's from main-queue. Internally, client will ack the message from main-queue once it's sent to retry queue.

> If i do `consumer.acknowledge(msg);` after sending to DLQ, does it remove from main queue or retry-queue?

Yes.


> 3\. We are processing millions of requests in a day, does this scale well?

It's ~10 message/second on average, quite small for pulsar actually, which can achieve 1 million msg/second.

**Reply by raunakagrawal47** [reply]
since I am manually sending the message to the DLQ after some retry based on message, how do I ack the message from the retry queue? (To remove it from retry queue and send it to DLQ) The consumer I initially created is for the main queue and I don't think it would work. Any thoughts on this?
Also, what if acknowledging msg fails, should we do negativeAck or reconsumeLater to replay the msg again?
@Jason918 @michaeljmarshall

**Reply by Jason918** [reply]
> The consumer I initially created is for the main queue and I don't think it would work. Any thoughts on this?

The default retry-queue is created and subscribed internally. So when you call `consumer.receive()`, you may receive message from retry-queue ,and you can ack the message like normal messages. Just try it out :)

> Also, what if acknowledging msg fails, should we do negativeAck or reconsumeLater to replay the msg again?

This depends on the failure reason. You can retry ack if it's network issue, or just log the error to see what's going on.
`negativeAck` or `reconsumeLater` seems not necessary in your case, since you either finish processed the message or have sent the message to your own DLQ.

**Reply by raunakagrawal47** [reply]
@Jason918 Do we need to acknowledge the current message everytime before we send a new message to the retry queue, using reconsumelater()? Else, how will current message get acked and removed from the retry queue?

**Reply by Jason918** [reply]
You don't need manually send message to retry queue. Technically, you don't need to be aware of the retry-queue at all. So you don't need to do anything else when calling `reconsumelater()`.

Internally, in `reconsumelater()` pulsar client will acked the message after it sends it to the retry-queue, regardless of it's from main-queue or retry-queue (start from 2nd retry),

**Reply by raunakagrawal47** [reply]
@Jason918 @michaeljmarshall facing below issue: https://github.com/apache/pulsar/issues/12021

I am trying to produce 10 messages at a time. And all these 10 messages should be getting retried 'x' number of times by 1 consumer. ('x' varies at message level). But it errors out and is not able to retry.
The error message says: `Need Non-Null value for key`. How can we resolve it?

**Reply by Jason918** [reply]
This issue is not verified on the latest pulsar version. 
Which pulsar version are you using?

**Reply by raunakagrawal47** [reply]
We are using broker (server) version `2.8` and pulsar-client version on the java application `2.10.0 `

**Reply by Jason918** [reply]
I suggest you can open an issue to describe this issue with more details, including broker version 2.8.x?, sample code to reproduce this.

**Reply by raunakagrawal47** [reply]
The issue got resolved. I upgraded it to 2.10.1 (the latest version) And basically, the issue was, I was setting some custom properties which had a key and a value, but the value was null, so it was not able to process. Thanks for the help @Jason918 :)
